### PR TITLE
Parameterized kubectl once and for all :)

### PR DESCRIPTION
**Pull request by jayunit100**
Parameterizes `kubectl.sh`.  
This is IMO the best solution, it may require a very minor change to your CI.

```
_output/local/bin/linux/amd64/e2e.test  \
-provider="local" --host="http://127.0.0.1:8080" \
-ginkgo.focus="kubectl*" \
-kubeconfig=/opt/kube-creds \
-KubectlPath="cluster/kubectl.sh" \
-repo-root=/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/
```

### CODE REVIEW

**Comment by jayunit100** [comment]
@ixdy @timothysc @rrati any  thoughts on this? It will allow us to run kubectl tests alot easier both for dev and CI .

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -82,6 +82,7 @@ func init() {
 	flag.StringVar(&testContext.Host, "host", "", "The host, or apiserver, to connect to")
 	flag.StringVar(&testContext.RepoRoot, "repo-root", "../../", "Root directory of kubernetes repository, for finding test files.")
 	flag.StringVar(&testContext.Provider, "provider", "", "The name of the Kubernetes provider (gce, gke, local, vagrant, etc.)")
+	flag.StringVar(&testContext.KubectlPath, "KubectlPath", "cluster/kubectl.sh", "The kubectl binary to use. For binaries, you might just use 'kubectl' for this arg.")
```
Comment (bgrant0607) [inline test/e2e/e2e_test.go]: Should be all lowercase, with a hyphen between words, like all other flags.

**Comment by bgrant0607** [comment]
Seems like a reasonable thing to do for now. Eventually I'd like to eliminate the need for a wrapper script -- #8817

**Comment by jayunit100** [comment]
okiedokie @bgrant0607 fixed the parameter and rebased ....  thanks for the review

**Comment by jlowdermilk** [comment]
lgtm

File: test/e2e/util.go [hunk 2]
```diff
@@ -481,9 +482,14 @@ func kubectlCmd(args ...string) *exec.Cmd {
 	}
 	kubectlArgs := append(defaultArgs, args...)
 
-	//TODO: the "kubectl" path string might be worth externalizing into an (optional) ginko arg.
-	cmd := exec.Command(filepath.Join(testContext.RepoRoot, "cluster/kubectl.sh"), kubectlArgs...)
-	Logf("Running '%s %s'", cmd.Path, strings.Join(cmd.Args, " "))
+	//Allow users to specify KubectlPath at commandline... so that we can test kubectl (the binary)
+	//Or so that, in the dev mode, we can test the kubectl shell script.
+	cmdToRun := testContext.KubectlPath
```
Comment (ixdy) [inline test/e2e/util.go]: I'd prefer this preserved the behavior of joining relative paths to testContext.RepoRoot.

Comment (jayunit100) [inline test/e2e/util.go]: hi ixdy.  

Good eye :) Here is my thinking: 

We *can* do RepoRoot append if  you are okay with an "empty" or "nil" default value for 
- KubectlPath which serves as an indicator that we need to dynamically generate the path from RepoRoot.
- current comments for RepoRoot indicate that its mainly for `examples` and this seemed more like a temporary solution. 

Im okay either way  , just wanted to make sure you realized this tradeoff .  let me know your thoughts and ill implement.

Comment (ixdy) [inline test/e2e/util.go]: I think the best thing might be to revert to using the kubectl binary as we did before my change, rather than passing through cluster/kubectl.sh.

So basically, change the parameter default to "kubectl" and I think we're probably good. (Only issue is that it might break Jenkins, if kubectl isn't properly in the `PATH` anymore.)

**Comment by jayunit100** [comment]
okay, now it uses `kubectl` as the default.   I tested the arg and it works for dev if you just pass `_output/local/bin/linux/amd64/e2e.test  -provider="local" --host="http://127.0.0.1:8080" -ginkgo.focus="kubectl*" --kubeconfig=/opt/kube-creds -repo-root=/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/ -kubectl-path=cluster/kubectl.sh`

File: test/e2e/e2e_test.go [hunk 3]
```diff
@@ -82,6 +82,7 @@ func init() {
 	flag.StringVar(&testContext.Host, "host", "", "The host, or apiserver, to connect to")
 	flag.StringVar(&testContext.RepoRoot, "repo-root", "../../", "Root directory of kubernetes repository, for finding test files.")
 	flag.StringVar(&testContext.Provider, "provider", "", "The name of the Kubernetes provider (gce, gke, local, vagrant, etc.)")
+	flag.StringVar(&testContext.KubectlPath, "kubectl-path", "kubectl", "The kubectl binary to use. For development, you might us 'cluster/kubectl.sh' here.")
```
Comment (ixdy) [inline test/e2e/e2e_test.go]: nit: s/us/use/

File: test/e2e/util.go (hunk 2 above)
Comment (ixdy) [inline test/e2e/util.go]: sorry for so many nits, but I don't think having cmdToRun as a separate variable really makes sense. just fold this into exec.Command below.

File: test/e2e/util.go [hunk 4]
```diff
@@ -481,9 +482,14 @@ func kubectlCmd(args ...string) *exec.Cmd {
...
+
+	//i.e. here we might define a command like "cluster/kubectl.sh get pods"
```
Comment (ixdy) [inline test/e2e/util.go]: this comment probably wants to be deleted, too.

**Comment by jayunit100** [comment]
okay pushed/updated the comments  ....

**Comment by ixdy** [comment]
I think I'd also feel a little more comfortable about going back to bare `kubectl` if `hack/ginkgo-e2e.sh` was updated to set the path as it used to do, e.g. like https://github.com/GoogleCloudPlatform/kubernetes/blob/f4bb6aa37d6dc3f244ad22204812255c6ecb148b/hack/ginkgo-e2e.sh#L120

**Comment by jayunit100** [comment]
ill find you on irc for the last batch :)

**Comment by jayunit100** [comment]
okay updated PR w/ the change to `hack`.  looks okay now?  Never used the `hack/ginkgo-e2e.sh` script so please give that change a good look !

File: hack/ginkgo-e2e.sh [hunk 5]
```diff
@@ -88,10 +88,13 @@ if [[ ${GINKGO_PARALLEL} =~ ^[yY]$ ]]; then
...
 fi
 
+
 # The --host setting is used only when providing --auth_config
 # If --kubeconfig is used, the host to use is retrieved from the .kubeconfig
 # file and the one provided with --host is ignored.
-"${ginkgo}" "${ginkgo_args[@]:+${ginkgo_args[@]}}" "${e2e_test}" -- \
+# Add path for things like running kubectl binary.
+export PATH=$(dirname "${e2e_test}"):"${PATH}"
+"${ginkgo}" "${ginkgo_args[@]:+${ginkgo_args[@]}}" "${e2e}" -- \
```
Comment (ixdy) [inline hack/ginkgo-e2e.sh]: please revert the change to this line. the variable is ${e2e_test}, not ${e2e}

**Comment by jayunit100** [comment]
okiedokie updated again

**Comment by ixdy** [comment]
LG, thanks!
